doc/decisions/0007-feature-toggles-development.md

```diff
@@ -0,0 +1,92 @@
+# 0007. Feature toggles development
+
+| Date | Tags |
+|---|---|
+| 2018-05-25 | development |
+
+
+## Status
+
+Proposed
+
+## Context
+
+This change describes a methodology of development which we believe will help
+us increase the confidence and speed in the development process.
+
+### What problem are we trying to solve
+
+Currently the time between a PR is sent and the code gets merged is between 1 day
+to in some case 4 to 5 days.
+
+This has some disadvantages:
+
+1) This workflows incentivise developers to make large PRs, as building 
+incrementally but having to wait 2 to 3 days for the code to be merged at each
+step is not desirable and results in loss of context.
+
+2) If a bug is introduced it is harder to fix, as the longer it takes to integrate code 
+the more likely the developer has lost familiarity with the code.
+
+3) Constant rebases introduce an overhead for the developer and can introduce
+subtle bugs, as rebased code is not thoroughly reviewed.
+
+4) Code review is only performed once, but during the manual QA step significant
+changes might be added, which are not reviewed anymore.
+
+### How are we going to address it
+
+We believe feature toggles development https://martinfowler.com/articles/feature-toggles.html
+is a step in the right direction.
+
+The way it would work is:
+
+A developer starts working on a new feature.
+Any new code will be scoped behind a feature flag and:
+
+1) no significant changes are made to the existing codebase
+2) no changes are visible to the user when the feature flag is off
```

Comment:
This sounds neat and simple, but really only works well when the change is isolated. While I agree that this is something we should strive for in general, its simply doesn't work when it isn't, i.e. in that case the logic behind the flag application would be so complex it could itself be a source of bugs.

Reply:
Yes, only some of the PRs will be able to get a free pass,  I think the majority will still go through manual QA, as mentioned

Reply:
We used a similar approach at Opera. One good thing about it that you catch integration bugs much earlier.
For example, you make just some groundwork for a feature and refactor something, with that approach you can submit a refactoring or partial change earlier and if anything is wrong with integration it will be uncovered early.
So here I wouldn't really strive for isolation of the code, but more for being able to PR a part of a feature, that is small, instead of submitting a week's worth of work at once.
More issues are uncovered during code reviews in that case too.